### PR TITLE
Improve BaseResolver Thrown Exceptions

### PR DESCRIPTION
**Pull request by broderickt2**
Add new JsonPatchOperationException and implement it in tests and BaseResolver. Add TestAdapter and TestFramework packages to use Assert.ThrowsException

### CODE REVIEW

File: src/JsonPatch.Tests/packages.config [hunk 1]
```diff
@@ -1,6 +1,8 @@
 ﻿<?xml version="1.0" encoding="utf-8"?>
 <packages>
   <package id="Microsoft.Bcl.AsyncInterfaces" version="7.0.0" targetFramework="net48" />
+  <package id="MSTest.TestAdapter" version="3.1.1" targetFramework="net48" />
```
Comment (broderickt2) [inline src/JsonPatch.Tests/packages.config:4]: Had to update the Tests project references as well by removing the below in my screenshot. It is not exactly showing up in the diff
image

File: src/JsonPatch.Common/JsonPatchOperationException.cs [hunk 2]
```diff
@@ -0,0 +1,59 @@
+﻿using System;
+
+namespace JsonPatch
+{
+    /// <summary>
+    /// Raised when an error occurs while processing changes. Includes extra details about what was being processed.
+    /// </summary>
+    public class JsonPatchOperationException : JsonPatchException
```
Comment (broderickt2) [inline src/JsonPatch.Common/JsonPatchOperationException.cs:8]: @myquay From a backwards compatibility standpoint, does it make sense to create this new exception type? Why not modify `JsonPatchException` and add a new constructor and properties to this class?

I just worry about a scenario where someone was catching this specific exception on their end during a patch operation and now it changes to `JsonPatchOperationException`.

**Comment by broderickt2** [comment]
Should `BaseResolver.GetValueFromPath` use `JsonPatchOperationException`?

File: src/JsonPatch.Common/JsonPatchOperationException.cs (hunk 2 above)
Comment (myquay) [reply src/JsonPatch.Common/JsonPatchOperationException.cs:8]: It should be backwards compatible. The catch block will always catch the most specific exception available to it, working its way up from the inheritance hierarchy - so a `catch(JsonPatchException ex)` will also catch the `JsonPatchOperationException`.

The main reason was to give flexibility around raising an exception in a context where it didn't make sense to return all the additional information.

**Comment by broderickt2** [comment]
@myquay I meant to ask. I found in a lot of the files there are unused references at the top of them. Are you aware of this or is this intentional? I am under the mindset that less code is easier to maintain. Are you against cleaning these up if created a pull request?
Here are 2 examples:
image

image

**Pull request merged**